lib/autoload/duplicator.rb

```diff
@@ -0,0 +1,26 @@
+class Duplicator
+  # Create an instance of Duplicator to track duplicated objects.
+  #
+  # @param [#to_set] excluded_objects An Enumerable of objects to be excluded from duplication
+  def initialize(excluded_objects = [])
+    @duplicated_objects = {}  # hash to check what has been duplicated
+    @exclusion_set = excluded_objects.to_set  # set to check what should be excluded
+  end
+
+  # Deep copy +source_object+ and its children. source_object must provide a duplicate
+  # method which duplicates its children.
+  #
+  # @param [#duplicate] source_object The object to be duplicated.
+  # @return duplicated_object A reference to the duplicated object.
+  def duplicate_object(source_object)
+    if @duplicated_objects.key?(source_object)
+      @duplicated_objects[source_object]
+    elsif !@exclusion_set.include?(source_object)
+      @duplicated_objects[source_object] = source_object.dup
```

Comment:
So I see why it works -- you use two-step construction. Any way to construct it in one step, like Object#dup? It calls initialize_copy.